osutil/group.go

```diff
@@ -88,7 +109,28 @@ var findUidNoGetentFallback = func(username string) (uint64, error) {
 var findGidNoGetentFallback = func(groupname string) (uint64, error) {
 	group, err := user.LookupGroup(groupname)
 	if err != nil {
-		return 0, err
+		// Treat all non-nil errors as user.Unknown{User,Group}Error's, as
+		// currently Go's handling of returned errno from get{pw,gr}nam_r
+		// in the cgo implementation of user.Lookup is lacking, and thus
+		// user.Unknown{User,Group}Error is returned only when errno is 0
+		// and the list of users/groups is empty, but as per the man page
+		// for get{pw,gr}nam_r, there are many other errno's that typical
+		// systems could return to indicate that the user/group wasn't
+		// found, however unfortunately the POSIX standard does not actually
+		// dictate what errno should be used to indicate "user/group not
+		// found", and so even if Go is more robust, it may not ever be
+		// fully robust. See from the man page:
+		//
+		// > It [POSIX.1-2001] does not call "not found" an error, hence
+		// > does not specify what value errno might have in this situation.
+		// > But that makes it impossible to recognize errors.
+		//
+		// See upstream Go issue: https://github.com/golang/go/issues/40334
+
+		// if there is a real problem finding the user/group then presumably
+		// other things will fail upon trying to create the user, etc. which
+		// will give more useful and specific errors
+		return 0, user.UnknownGroupError(groupname)
```

Comment:
I'm not thrilled the the user id being returned in an error case being the all-powerful account on the system. Are there any "sum types" in golang that could be used to clearly differentiate between errors and successes, that would make failing open more difficult?

Reply:
it seems to me that there are kinda 2 points you're making here which I will respond to separately

1) yes there is a concept of "sum types", specifically for this case of errors, there is a new system for composing different error types together and being able to inspect the full set of composed error types, which would allow us to return both the original error we got from user.Lookup here, as well as an error of the type user.UnknownGroupError that can be inspected / checked for explicitly by calling code, but I would consider that kind of change a bit invasive for this PR and I think is more a stylistic choice here rather than necessary.
2) regarding returning the default value of 0 here, this is fairly idiomatic Go, which is just returning the default return value for a function when there is an error, however in this case it is rather unfortunate that the default value of a uint also happens to be a fairly powerful user in the success case. What we could do to avoid this and protect against ignoring the error is we could change the return value on error to something that's not idiomatic Go, and return a very large number instead of 0, or we could change the return type to be a pointer, and on error we return nil, which means that a caller that ignores the error would panic trying to use the value as it would be nil. The issue with the former is that it's unclear to me what the right value is, looking around it seems different OS's have different max uid's, and there's not an easy way to find out without adding a bunch of OS specific code we would rather not do (but probably technically cloud). The issue with the latter is that we would have to change the function signature and all usages, which for this PR is rather invasive, so I would prefer to do that in a followup PR if you think this change needs to be made. I would also have to audit all other uid returning code we have to see if there are other functions similarly affected by this bug of default return value also meaning something very powerful.

Reply:
It seems this has a single +1 now but no security review +1. Would changing the return value in the error case of "0" to "-1" be sufficient so that @setharnold would do a +1 on this? Or is there more needed?

Reply:
-1 has also a special meaning for some syscalls, so not sure it's a completely better solution

Reply:
I think of the choices between 0 and -1, I prefer 0, especially if that's the convention.

I would prefer to see a move to using sum types that can clearly prevent using an error return as a legitimate value but can appreciate that might be a long-term project.

Security team +1 on this. (Jamie's otherwise occupied; if you've got a specific question for him you can ask him more directly.)

Thanks

Reply:
let's add a TODO to this file to turn the uint results at a later point to something with a distinguished default value (i.e. that can't be confused with uid/gid 0)

Reply:
done in #9243